R/combineResults.R

```diff
@@ -0,0 +1,197 @@
+#'     method="cluster", clusters=test$label) 
+#' pred4 <- SingleR(test, ref2, labels=ref2$label, 
+#'     method="cluster", clusters=test$label) 
+#'
+#' ###############################
+#' ##     Combining results     ##
+#' ###############################
+#'
+#' pred.single <- combineResults(list("pred1" = pred1, "pred2" = pred2))
+#' pred.clust <- combineResults(list("pred3" = pred3, "pred4" = pred4))
+#'
+#' @seealso
+#' \code{\link{matchReferences}}, to harmonize labels between reference datasets.
+#' \code{\link{SingleR}}, for generating predictions.
+#'
+#' @importFrom S4Vectors DataFrame metadata metadata<-
+#'
+#' @export
+combineResults <- function(results) {
+    if (length(unique(lapply(results, rownames))) != 1) {
+        stop("cell/cluster names are not identical")
+    }
+
+    if (length(unique(lapply(results, function (x) sort(metadata(x)$common.genes)))) != 1) {
+        # This should be changed to 'stop' before release/after merge with PR #60.
+        warning("common genes are not identical")
+    }
+
+    has.first <- FALSE
+    has.pruned <- FALSE
+    has.de <- FALSE
+
+    if (!is.null(results[[1]]$first.labels)) {
+        has.first <- TRUE
+    }
+
+    if (!is.null(results[[1]]$pruned.labels)) {
+        has.pruned <- TRUE
+    }
+
+    if (!is.null(metadata(results[[1]])$de.genes)) {
+        has.de <- TRUE
+        de.names <- sapply(results, function(x) names(metadata(x)$de.genes))
+    }
+
+    ncells <- nrow(results[[1]])
+
+    last.best <- rep(-Inf, ncells) 
+    chosen.label <- chosen.first <- chosen.pruned <- rep(NA_character_, ncells)
+    collected.scores <- list()
+    collected.de <- list()
+
+    for (i in seq_along(results)) {
+        res <- results[[i]]
+        scores <- res$scores
+        curbest <- scores[cbind(seq_len(ncells), max.col(scores))]
+        collected.scores[[i]] <- scores
+        better <- curbest > last.best
+         
+        chosen.label[better] <- res$labels[better]
+
+        if (has.first) { # either everyone has 'first', or no-one does.
+            chosen.first[better] <- res$first.labels[better]
+        }
+
+        if (has.pruned) { # same for pruned.
+            chosen.pruned[better] <- res$pruned.labels[better]
+        }
+
+        if (has.de) {
+            collected.de[[i]] <- metadata(res)$de.genes
+        }
+
+        last.best <- curbest
+    }
+
+    all.scores <- do.call(cbind, collected.scores)
+
+    output <- DataFrame(scores = I(all.scores), labels = chosen.label)
+    metadata(output)$common.genes <- metadata(results[[1]])$common.genes
+
+    if (has.de) {
+        metadata(output)$de.genes <- do.call(cbind, collected.de)
+        names(metadata(output)$de.genes) <- de.names
+    }
+
+    if (has.first) {
+        output$first.labels <- chosen.first
+    }
+
+    if (has.pruned) {
+        output$first.labels <- chosen.pruned
+    }
+
+    output$orig.results <- do.call(DataFrame, lapply(results, I))
+
+    return(output)
```

Comment:
As a general rule, no need for `return` at the end of a function; in R code, I only use `return` when I'm breaking out of the function prematurely. More of a coding style thing but I think that's what we do in the rest of the package.